src/oracle/middleware.js

```diff
@@ -0,0 +1,805 @@
+  assert(CREDENTIALS_FILE != "", '$CREDENTIALS_FILE is required');
+  assert(OFFERS_FILE != "", '$OFFERS_FILE is required');
+  assert(FEEDS_FILE != "", '$FEEDS_FILE is required');
+}
+
+var agoricNames = {};
+var fromBoard = {};
+var vstorage = {};
+const marshaller = boardSlottingMarshaller();
+
+var credentials;
+var feeds;
+
+/**
+  * Function to load from file or initialise it
+  * @returns empty state or state from file
+  */
+const readState = () => {
+  // try to read JSON file and if it doesnt exist, create one
+  try {
+    //read JSON file
+    return readJSONFile(STATE_FILE)
+  } catch (err) {
+
+    //create an initial state
+    let initial_state = {
+      "jobs": [],
+      "previous_results": {}
+    }
+
+    //save state
+    saveJSONDataToFile(initial_state, STATE_FILE)
+    return initial_state
+  }
+}
+
+/**
+  * Function to initialise state
+  */
+const initialiseState = () => {
+  //read state
+  let state = readState()
+
+  //go through each job
+  for (let index in state.jobs) {
+    let currentJob = state.jobs[index]
+    //if not initialised
+    if (!(currentJob.name in state.previous_results)) {
+      state.previous_results[currentJob.name] = {
+        id: currentJob.job,
+        result: 0,
+        round: {}
+      }
+    }
+  }
+
+  //save state
+  saveJSONDataToFile(state, STATE_FILE);
+}
+
+/**
+  * Function to get the job's index
+  * @param {*} jobName the job name 
+  * @returns the index of the job
+  */
+const getJobIndex = (jobName) => {
+  //read state
+  let state = readState()
+
+  //go through each job
+  for (let index in state.jobs) {
+    let currentJob = state.jobs[index]
+    //if not initialised
+    if (currentJob.name == jobName) {
+      return index
+    }
+  }
+  return -1
+}
+
+
+/**
+  * Function to send a job run to the CL node
+  * @param {*} credentials the external initiator credentials
+  * @param {*} count the request id
+  * @param {*} jobId the Chainlink external job id
+  * @param {*} chainlinkUrl the Chainlink node url where to send the job request
+  * @param {*} requestType the request type, 1 = time, 2 = deviation, 3 = new round
+  */
+const sendJobRun = async (credentials, count, jobId, chainlinkUrl, requestType) => {
+  const options = {
+    url: chainlinkUrl + "/v2/jobs/" + jobId + "/runs",
+    body: {
+      "payment": 0,
+      "request_id": count,
+      "request_type": requestType
+    },
+    headers: {
+      "Content-Type": "application/json",
+      "X-Chainlink-EA-AccessKey": credentials["EI_IC_ACCESSKEY"],
+      "X-Chainlink-EA-Secret": credentials["EI_IC_SECRET"]
+    },
+    method: 'POST',
+  };
+
+  //try request with loop retries
+  let error = ""
+  for (let i = 0; i < SUBMIT_RETRIES; i++) {
+    try {
+      let res = await axios.post(options.url, options.body, {
+        timeout: 60000,
+        proxy: false,
+        headers: options.headers,
+        httpAgent: new http.Agent({ keepAlive: true })
```

Comment:
be careful of this. Like this you are saying that after the request is served, the connection should remain open, therefore, you will swamp the middleware and the node with a lot of connections.
```suggestion
        httpAgent: new http.Agent({ keepAlive: false })
```

Reply:
Fixed in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/47346bdfb01465b4ef33fca7ce6b0ee0a0d94f05#diff-d72ebe1db3542d79d3f983a81b4b5a78ebdca8897dec342d7403f3fa0f0f3f69